tests/codex/slots/backends/testcircomcompat.nim

```diff
@@ -0,0 +1,137 @@
+import pkg/codex/stores
+
+import pkg/constantine/math/arithmetic
+import pkg/constantine/math/io/io_fields
+import pkg/constantine/math/io/io_bigints
+
+import ./helpers
+import ../helpers
+
+suite "Test Circom Compat Backend - control inputs":
+  let
+    r1cs = "tests/circuits/fixtures/proof_main.r1cs"
+    wasm = "tests/circuits/fixtures/proof_main.wasm"
+
+  var
+    circom: CircomCompat
+    verifyingKeyPtr: ptr CircomKey
+    proofInput: ProofInput[Poseidon2Hash]
+    publicInputs: CircomInputs
+
+  setup:
+    let
+      inputData = readFile("tests/circuits/fixtures/input.json")
+      inputJson = parseJson(inputData)
+
+    proofInput = jsonToProofInput[Poseidon2Hash](inputJson)
+    publicInputs = toPublicInputs[Poseidon2Hash](proofInput).toCircomInputs
+
+    # circom = CircomCompat.init(r1cs, wasm, zkey)
+    circom = CircomCompat.init(r1cs, wasm)
+    verifyingKeyPtr = circom.getVerifyingKey().tryGet
+
+  teardown:
+    publicInputs.releaseNimInputs()      # this is allocated by nim
+    verifyingKeyPtr.addr.releaseKey()    # this comes from the rust FFI
+    circom.release()                     # this comes from the rust FFI
+
+  test "Should verify with correct inputs":
+    let
+      proof = circom.prove(proofInput).tryGet
+
+    check circom.verify(proof, publicInputs, verifyingKeyPtr[]).tryGet
+
+  test "Should not verify with incorrect inputs":
+    proofInput.slotIndex = 1 # change slot index
+
+    let
+      proof = circom.prove(proofInput).tryGet
+
+    check circom.verify(proof, publicInputs, verifyingKeyPtr[]).tryGet == false
+
+suite "Test Circom Compat Backend":
+  let
+    slotId = 3
+    samples = 5
+    ecK = 2
+    ecM = 2
+    numDatasetBlocks = 8
+    blockSize = DefaultBlockSize
+    cellSize = DefaultCellSize
+
+    r1cs = "tests/circuits/fixtures/proof_main.r1cs"
+    wasm = "tests/circuits/fixtures/proof_main.wasm"
+
+  var
+    store: BlockStore
+    manifest: Manifest
+    protected: Manifest
+    verifiable: Manifest
+    circom: CircomCompat
+    verifyingKeyPtr: ptr CircomKey
+    proofInput: ProofInput[Poseidon2Hash]
+    publicInputs: CircomInputs
+    challenge: array[32, byte]
+    builder: Poseidon2Builder
+    sampler: Poseidon2Sampler
+
+  setup:
+    let
+      repoDs = SQLiteDatastore.new(Memory).tryGet()
+      metaDs = SQLiteDatastore.new(Memory).tryGet()
+
+    store = RepoStore.new(repoDs, metaDs)
+
+    (manifest, protected, verifiable) =
+        await createVerifiableManifest(
+          store,
+          numDatasetBlocks,
+          ecK, ecM,
+          blockSize,
+          cellSize)
+
+    builder = Poseidon2Builder.new(store, verifiable).tryGet
+    sampler = Poseidon2Sampler.new(slotId, store, builder).tryGet
+
+    # circom = CircomCompat.init(r1cs, wasm, zkey)
+    circom = CircomCompat.init(r1cs, wasm)
+    verifyingKeyPtr = circom.getVerifyingKey().tryGet
+    challenge = 1234567.toF.toBytes.toArray32
+
+    proofInput = (await sampler.getProofInput(challenge, samples)).tryGet
```

Comment:
I'm not sure why we have this suite in addition to the above, but I'm assuming this is testing generation of proof inputs. If that's the case, I'd probably would have preferred to have those tests on their own (i.e. generating proof inputs and assertions on generated proof inputs) as smaller granularity tests make tracking down bugs simpler.

Reply:
This is testing again control inputs that we test the circuit itself, this ensures that the circuit is instantiated properly in case we start seeing issues with the sampled chunks. This can probably go away at some point, but it makes me feel a little better having this test in there for now.

Reply:
btw, we do have more granular tests in `sampler/testsampler.nim`, which test the sampling itself rather than the circuit invocation.